### PR TITLE
graph: backend: dnnl: kernels: add decomp kernel for sdpa training fwd

### PR DESCRIPTION
**Pull request by TaoLv**
Optimize SDPA training forward (with softmax stats output) with the primitive decomposition kernel.

Support status as below. Fallback to larger_partition_kernel_t for unsupported cases.

- Platform: CPU only, with OMP, thread pool, TBB runtimes.
- Data type: f32, bf16, f16. (no quantized data types).
- Softmax requires stats output.
- No attention mask.
- No softmax `inf_as_zero` mode.
- No dropout.
- No GQA.

Fixes MFDNN-14757

### CODE REVIEW

**Comment by TaoLv** [comment]
make test
disable benchdnn_all
enable benchdnn_graph

**Review by rongzha1 — APPROVED** [review]

**Comment by TaoLv** [comment]
make test
disable benchdnn_all
enable benchdnn_graph

File: src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            = [= COMPAT_THIS_CAPTURE](int tid, int nthr, dim_t bo, dim_t bi) {
+#if DNNL_CPU_RUNTIME == DNNL_RUNTIME_THREADPOOL
+        threadpool_utils::deactivate_threadpool();
+#endif
+
+        // Prepare execution args and allocate real memory
+        prepare_sub_args(var_grantor, tid, block_size, res->mem_map);
+
+        const size_t group_head = sdp_cfg_.num_head_q / sdp_cfg_.num_head_kv;
+        const size_t wei_head_offset = bi / group_head;
```
Comment (vpirogov) [inline src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp]: Why this is needed is GQA is not supported?

File: src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        dnnl_primitive_execute_without_tp_hook(sdp_cfg_.sub_softmax_prim,
+                p_stream, res->sub_softmax_args[tid]);
+
+        // Compute stats: logsumexp = reduce_max(scores) - log(reduce_max(P))
+        {
+            auto &sub_stats_tid = res->mem_map[sdp_cfg_.sub_stats.get()][tid];
+            sub_stats_tid.set_data_handle(stats_user_pointer
+                    + (bo * sdp_cfg_.stats_dst_strides[0]
+                              + bi * sdp_cfg_.stats_dst_strides[1])
+                            * sizeof(float));
```
Comment (vpirogov) [inline src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp]: Do we check that stats tensor is dense somewhere?

File: src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp (hunk 1 above)
Comment (TaoLv) [reply src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp]: It's some legacy code inherited from the original sdp_dcomp.cpp. It's not needed and removed. I will add them back in a follow-up PR for adding GQA support.

File: src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp (hunk 2 above)
Comment (TaoLv) [reply src/graph/backend/dnnl/kernels/sdp_decomp_training.cpp]: Good catch! Added the check for stats format. An reorder will be called if the user requires stats output to be non-dense (inner strides are not (1, 1)). It's not needed for the interested case requested by customer.

The decomposed primitive sequence looks like below:
```
// qk matmul: bf16bf16f32
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx2_vnni_2,undef,src:bf16::blocked:ab::f0 wei:bf16::blocked:ba::f0 dst:f32::blocked:ab::f0,attr-scratchpad:user attr-post-ops:binary_mul:f32:0,,197x64:64x197,0.0380859
// softmax - f32 prob
onednn_verbose,v1,primitive,exec,cpu,softmax,jit:avx2_vnni_2,forward_inference,src:f32::blocked:ab::f0 dst:f32::blocked:ab::f0,attr-scratchpad:user,alg:softmax_accurate axis:1,197x197,0.0300293
// reduction 1 for stats output
onednn_verbose,v1,primitive,exec,cpu,reduction,jit:avx2_vnni_2,undef,src:f32::blocked:ab::f0 dst:f32::blocked:ab::f0,attr-scratchpad:user attr-post-ops:eltwise_log,alg:reduction_max p:0 eps:0,197x197:197x1,0.0498047
// reduction 2 for stats output
onednn_verbose,v1,primitive,exec,cpu,reduction,jit:avx2_vnni_2,undef,src:f32::blocked:ab::f0 dst:f32::blocked:ab::f0,attr-scratchpad:user attr-post-ops:binary_sub:f32:1,alg:reduction_max p:0 eps:0,197x197:197x1,0.0490723
// reorder stats from dense (1, 1) to user-specified strides (2, 1)
onednn_verbose,v1,primitive,exec,cpu,reorder,jit:uni,undef,src:f32::blocked:ab::f0 dst:f32::blocked:ab:2x1:f0,attr-scratchpad:user,,197x1,0.000976562
// reorder f32 prob to bf16
onednn_verbose,v1,primitive,exec,cpu,reorder,jit:uni,undef,src:f32::blocked:ab::f0 dst:bf16::blocked:ab::f0,attr-scratchpad:user,,197x197,0.00317383
// pv matmul: bf16bf16bf16, sdpa output.
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx2_vnni_2,undef,src:bf16::blocked:ab::f0 wei:bf16::blocked:ab::f0 dst:bf16::blocked:ab::f0,attr-scratchpad:user,,197x197:197x64,0.0380859
```

**Review by vpirogov — APPROVED** [review]

**Comment by TaoLv** [comment]
make test
disable benchdnn_all
enable benchdnn_graph

**Comment by TaoLv** [comment]
make test
disable benchdnn_all
enable benchdnn_graph
